pxr/base/tf/delegatedCountPtr.h

```diff
@@ -0,0 +1,268 @@
+//
+// Copyright 2023 Pixar
+//
+// Licensed under the Apache License, Version 2.0 (the "Apache License")
+// with the following modification; you may not use this file except in
+// compliance with the Apache License and the following modification to it:
+// Section 6. Trademarks. is deleted and replaced with:
+//
+// 6. Trademarks. This License does not grant permission to use the trade
+//    names, trademarks, service marks, or product names of the Licensor
+//    and its affiliates, except as required to comply with Section 4(c) of
+//    the License and to reproduce the content of the NOTICE file.
+//
+// You may obtain a copy of the Apache License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the Apache License with the above modification is
+// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. See the Apache License for the specific
+// language governing permissions and limitations under the Apache License.
+//
+
+#ifndef PXR_BASE_TF_DELEGATED_COUNT_PTR_H
+#define PXR_BASE_TF_DELEGATED_COUNT_PTR_H
+
+#include "pxr/pxr.h"
+#include "pxr/base/tf/tf.h"
+#include "pxr/base/tf/api.h"
+
+#include "pxr/base/tf/diagnosticLite.h"
+
+#include <memory>
+#include <type_traits>
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+/// When constructing a `TfDelegatedCountPtr` from a raw pointer, use the
+/// `TfDelegatedCountIncrementTag` to explicitly signal that the pointer's
+/// delegated count should be incremented on construction. This is the most
+/// common tag.
+constexpr struct TfDelegatedCountIncrementTagType{}
+    TfDelegatedCountIncrementTag{};
+
+/// When constructing a `TfDelegatedCountPtr` from a raw pointer, use the
+/// `TfDelegatedCountDoNotIncrementTag` to avoid incrementing the delegated
+/// count on construction. This must be carefully used to avoid memory errors.
+constexpr struct TfDelegatedCountDoNotIncrementTagType{}
+    TfDelegatedCountDoNotIncrementTag{};
+
+/// Stores a pointer to a `ValueType` which uses `TfDelegatedCountIncrement`
+/// and `TfDelegatedCountDecrement` to bookkeep.
+///
+/// The `TfDelegatedCountPtr` is responsible for calling
+/// `TfDelegatedCountIncrement` and `TfDelegatedCountDecrement` as needed in 
+/// construction, assignment, and destruction operations.
+///
+/// `TfDelegatedCountIncrement` and `TfDelegatedCountDecrement` are expected to
+/// be `noexcept`. The increment and decrement methods can assume pointers are
```

Comment:
Requiring these to be `noexcept` seems potentially too restrictive to me...  What's the motivation for this requirement?  Could we just transparently propagate the `noexcept`-ness of these so that it's a "pay for what you use" type of thing?